packages/dev-server/src/utils.ts

```diff
@@ -0,0 +1,86 @@
+/**
+ * Safely parse the URL path from a given string.
+ * Returns the pathname if parsing is successful, otherwise returns undefined.
+ *
+ * @example
+ * safeParseUrlPath('/foo/bar?query=123') // returns '/foo/bar'
+ */
+export const safeParseUrlPath = (value: string): string | undefined => {
+  try {
+    // Use dummy scheme host to parse relative URLs.
+    return new URL(value, 'http://localhost').pathname
+  } catch {
+    return undefined
+  }
+}
+
+/**
+ * Normalize the base path by ensuring it starts with a leading slash and does not end with a trailing slash.
+ * If the base is empty, '/', or undefined, it returns an empty string.
+ *
+ * @example
+ * normalizeBasePath('/foo/bar/') // returns '/foo/bar'
+ * normalizeBasePath('foo/bar') // returns '/foo/bar'
+ * normalizeBasePath('/') // returns ''
+ * normalizeBasePath('') // returns ''
+ * normalizeBasePath(undefined) // returns ''
+ */
+export const normalizeBasePath = (base: string | undefined): string => {
```

Comment:
The following wil return `''`.

```ts
normalizeBasePath('//')
```

Is this expected behavior? I think both are okay, but if so, it's better to add a test for it to show the intention.